desktop/notification/notify.go

```diff
@@ -0,0 +1,166 @@
+//
+// Message must include a summary and should include a body. The body may use
+// HTML-like markup language to include bold, italic or underline text, as well
+// as to include images and hyperlinks.
+//
+// A notification can automatically expire after the given number of
+// milliseconds. This is separate from the notification being visible or
+// invisible on-screen. Expired notifications are removed from persistent
+// message roster, if one is supported.
+//
+// A notification may replace an existing notification by setting the ReplacesID
+// to a non-zero value. This only works if the notification server was not
+// re-started and should be used for as long as the sender process is alive, as
+// sending the identifier across session startup boundary has no chance of
+// working correctly.
+//
+// A notification may optionally carry a number of hints that further customize it
+// in a specific way. Refer to various hint constructors for details.
+//
+// A notification may optionally also carry one of several actions. If
+// supported, actions can be invoked by the user, broadcasting a notification
+// response back to the session. This mechanism only works if there is someone
+// listening for the action being triggered.
+//
+// In all cases, the specific notification must take into account the
+// capabilities of the server. For instance, if a server does not support body
+// markup, then such markup is not automatically stripped by either the client
+// or the server.
+type Message struct {
+	AppName       string
+	Icon          string
+	Summary       string
+	Body          string
+	ExpireTimeout int32 // XXX: consider using time.Duration instead?
+	ReplacesID    ID
+	Actions       []Action
+	Hints         []Hint
+}
+
+// ID is the opaque identifier of a notification assigned by the server.
+//
+// Notifications with known identifiers can be closed or updated. The identifier
+// is valid within one desktop session and should not be used unless the calling
+// process initially sent the message.
+type ID uint32
+
+// Action describes a single notification action.
+//
+// ActionKey is returned in an D-Bus signal when an action is activated by the
+// user. The text must be localized for the appropriate language.
+type Action struct {
+	ActionKey     string
+	LocalizedText string
+}
+
+// Hint describes supplementeary information that may be used by the server.
+//
+// Various helpers create hint objects of specifc purpose.
+type Hint struct {
+	Name  string
+	Value interface{}
+}
+
+// ServerCapability describes a single capability of the notification server.
+//
+// Each server offers specific capabilities. It is advised to provide graceful
+// degradation of functionality, depending on the supported capabilities, so
+// that the notification messages are useful on a wide range of desktop
+// environments.
+type ServerCapability string
+
+// Server is an interface to a notification server interactions.
+type Server interface {
```

Comment:
Why use the interface here? Could we YAGNI it?

Reply:
Because for testing it's useful to give something a different implementation than the real server. Also because, although this is not finished yet, because we may have a way to have the extended GNOME notifications behind the same interface.